src/ipa/rpi/controller/rpi/awb_nn.cpp

```diff
@@ -0,0 +1,505 @@
+
+using namespace libcamera;
+
+LOG_DECLARE_CATEGORY(RPiAwb)
+
+constexpr double kDefaultCT = 4500.0;
+
+#define NAME "rpi.nn.awb"
+
+namespace RPiController {
+
+struct AwbNNConfig {
+	AwbNNConfig() {}
+	int read(const libcamera::YamlObject &params, AwbConfig &config);
+
+	/* An empty model will check default locations for model.tflite */
+	std::string model;
+	int numOutputs;
+	float minTemp;
+	float maxTemp;
+
+	bool enable_nn;
+
+	/* CCM matrix for 5000K temperature */
+	double ccm[9];
+};
+
+class AwbNN : public Awb
+{
+public:
+	AwbNN(Controller *controller = NULL);
+	~AwbNN();
+	char const *name() const override;
+	void initialise() override;
+	int read(const libcamera::YamlObject &params) override;
+
+protected:
+	void doAwb() override;
+
+private:
+	bool isAutoEnabled() const;
+	AwbNNConfig nnConfig_;
+	void transverseSearch(double t, double &r, double &b);
+	RGB processZone(RGB zone, float lux, float red_gain, float blue_gain);
+	void awbNN();
+	void loadModel();
+
+	std::unique_ptr<tflite::FlatBufferModel> model_;
+	std::unique_ptr<tflite::Interpreter> interpreter_;
+};
+
+int AwbNNConfig::read(const libcamera::YamlObject &params, AwbConfig &config)
+{
+	model = params["model"].get<std::string>("");
+	numOutputs = params["num_outputs"].get<int>(10);
+	minTemp = params["min_temp"].get<float>(2000.0);
+	maxTemp = params["max_temp"].get<float>(8000.0);
+
+	for (int i = 0; i < 9; i++)
+		ccm[i] = params["ccm"][i].get<double>(0.0);
+
+	enable_nn = params["enable_nn"].get<int>(1);
+
+	if (enable_nn) {
+		if (!config.hasCtCurve()) {
+			LOG(RPiAwb, Error) << "CT curve not specified";
+			enable_nn = false;
+		}
+
+		if (!model.empty() && model.find(".tflite") == std::string::npos) {
+			LOG(RPiAwb, Error) << "Model must be a .tflite file";
+			enable_nn = false;
+		}
+
+		bool valid_ccm = true;
+		for (int i = 0; i < 9; i++)
+			if (ccm[i] == 0.0)
+				valid_ccm = false;
+
+		if (!valid_ccm) {
+			LOG(RPiAwb, Error) << "CCM not specified or invalid";
+			enable_nn = false;
+		}
+
+		if (!enable_nn) {
+			LOG(RPiAwb, Warning) << "Neural Network AWB mis-configured - switch to Grey method";
+		}
+	}
+
+	if (!enable_nn) {
+		config.sensitivityR = config.sensitivityB = 1.0;
+		config.greyWorld = false;
+	}
+
+	return 0;
+}
+
+AwbNN::AwbNN(Controller *controller)
+	: Awb(controller)
+{
+}
+
+AwbNN::~AwbNN()
+{
+}
+
+char const *AwbNN::name() const
+{
+	return NAME;
+}
+
+int AwbNN::read(const libcamera::YamlObject &params)
+{
+	int ret;
+
+	ret = config_.read(params);
+	if (ret)
+		return ret;
+
+	ret = nnConfig_.read(params, config_);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+void AwbNN::loadModel()
+{
+	LOG(RPiAwb, Debug) << "Loading AWB model " << nnConfig_.model;
+	std::string modelPath = "/ipa/rpi/pisp/awb_model.tflite";
+	if (nnConfig_.model.empty()) {
+		std::string root = utils::libcameraSourcePath();
+		if (!root.empty()) {
+			modelPath = root + modelPath;
+		} else {
+			modelPath = LIBCAMERA_DATA_DIR + modelPath;
+		}
```

Comment:
Might be worth printing out the model path in an Info statement. People waste so much time thinking they're loading something that they aren't (myself included)!

Reply:
Ah I see we do it once it's loaded. Well, myabe a Debug statement then, so that folks can double-check if it's not being found.